<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Presentación grupo 5 : Rbims - dbCAN</title>
    <meta charset="utf-8" />
    <meta name="date" content="2022-08-05" />
    <script src="Presentación_files/header-attrs/header-attrs.js"></script>
    <link href="Presentación_files/remark-css/default.css" rel="stylesheet" />
    <link href="Presentación_files/remark-css/metropolis-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Presentación grupo 5 : Rbims - dbCAN
]
.subtitle[
## Nuevo workflow para trabajar con datos provenientes de dbCAN
]
.date[
### 2022-08-05
]

---

&gt; ####Roberto
&gt; ####Nelly 
&gt; ####Ulises
&gt; ####Julián
&gt; ####Andrés
&gt; ####Stephanie
&gt; ####Mirna y Joselyn

---

background-image: url(https://raw.githubusercontent.com/mirnavazquez/RbiMs/main/man/figures/Logo-rRbiMs.png)
background-position: center
background-size: contain




---
background-image: url(../Images/imagen.png)
background-position: center
background-size: contain

---
class: inverse, center, middle

# ¿Qué es dbCAN?


---
class: center, middle

# dbCAN 
## automated carbohydrate-active enzyme annotation 

Yanbin Yin, Xizeng Mao, Jincai Yang, Xin Chen, Fenglou Mao, Ying Xu, dbCAN: a web resource for automated carbohydrate-active enzyme annotation, Nucleic Acids Research, Volume 40, Issue W1, 1 July 2012, Pages W445–W451, [Liga al artículo](https://doi.org/10.1093/nar/gks479)

dbCAN es un servidor web para la anotación automática de enzimas activas de carbohidratos, financiado por la NFS. Recursos similares en la web incluyen CAZy, CAT (obsoleto) y dbCAN-sub. 

---
class: center, middle

# dbCAN 
## automated carbohydrate-active enzyme annotation 


El metaservidor dbCAN2 es una versión actualizada del servidor web dbCAN original.

El programa que ocupamo 'run_dbcan3' es es una herramienta que utiliza este servidor creado en su versión conda reescrito por Huang Lee; la versión V1 version fue escrita por  Tanner Yohe
---
## ¿Qué (intentamos) hacer?
## Creación de un paquete en GitHub 

https://github.com/Steph0522/Rbims_workflow  

####  Estructura (similar a Rbims)

.pull-left[

&gt; ####data

&gt; ####docs

&gt; ####tests

&gt; #### vignettes

&gt; README.Rmd

&gt; ...


]

.pull-right[

&gt; ####R (se agregaron dos nuevas función)

  `read_dbcan` (2022_08_05-READ-read_dbCAN.R)
  `bubble_dbcan` (2022_08_05-READ-read_dbCAN.R)

    
   Objetivos:
   
       Leer output de dbcan
       Filtrar nombres y señal de bins
       Calcular estadísticas
       Graficar un bubble plot 
]



---
# Instalación de dbCAN y base de datos

Instala  **run_dbcan3** de [Github](https://github.com/linnabrown/run_dbcan):

```{.sh}
conda create -n run_dbcan python=3.8 dbcan -c conda-forge -c bioconda
conda activate run_dbcan 
```

La base de datos:  
```{.sh}
test -d db || mkdir db
cd db \
    &amp;&amp; wget http://bcb.unl.edu/dbCAN2/download/CAZyDB.09242021.fa &amp;&amp; diamond makedb --in CAZyDB.09242021.fa -d CAZy \
    &amp;&amp; wget https://bcb.unl.edu/dbCAN2/download/Databases/V10/dbCAN-HMMdb-V10.txt &amp;&amp; mv dbCAN-HMMdb-V10.txt dbCAN.txt &amp;&amp; hmmpress dbCAN.txt \
    &amp;&amp; wget http://bcb.unl.edu/dbCAN2/download/Databases/tcdb.fa &amp;&amp; diamond makedb --in tcdb.fa -d tcdb \
    &amp;&amp; wget http://bcb.unl.edu/dbCAN2/download/Databases/tf-1.hmm &amp;&amp; hmmpress tf-1.hmm \
    &amp;&amp; wget http://bcb.unl.edu/dbCAN2/download/Databases/tf-2.hmm &amp;&amp; hmmpress tf-2.hmm \
    &amp;&amp; wget http://bcb.unl.edu/dbCAN2/download/Databases/stp.hmm &amp;&amp; hmmpress stp.hmm \
    &amp;&amp; cd ../ &amp;&amp; wget http://bcb.unl.edu/dbCAN2/download/Samples/EscheriaColiK12MG1655.fna \
    &amp;&amp; wget http://bcb.unl.edu/dbCAN2/download/Samples/EscheriaColiK12MG1655.faa \
    &amp;&amp; wget http://bcb.unl.edu/dbCAN2/download/Samples/EscheriaColiK12MG1655.gff
```

---
# Instalación de SignalP (Signalp Petitide prediction)
```{.bash}
mkdir -p run_dbcan/tools &amp;&amp; run_dbcan/tools/
tar -xvf signalp-4.1g.Linux.tar.gz &amp;&amp; cd signalp-4.1
```

# Corriendo dbcan

```{.bash}
for x in $(ls /home/user04/07.Bins/Proteoma/*.faa | cut -d "_" -f 1-3 | cut -d "/" -f 6); do mkdir ${x} ; done

for x in $(ls /home/user04/07.Bins/Proteoma/*.faa | cut -d "_" -f 1-3 | cut -d "/" -f 6); do run_dbcan /home/user04/07.Bins/Proteoma/${x}_renamed.fa.faa protein --out_dir /home/megadriveCURSO/user04/15.test_dbCan/${x}/ --db_dir /home/mirna/db/ --use_signalP=TRUE -sp /home/megadriveCURSO/mirna/run_dbcan/tools/signalp-4.1/signalp; done
```
---
class: inverse, center, middle

# Haciendo las funciones

---

## Funcion de lectura de archivo

```r
read_dbcan&lt;-function(dbcan_path){
  ruta_dbcan&lt;-dbcan_path
  # Load all the data tables results ---------------------------------------####
  lapply_read_delim_bind_rows &lt;- function(path, pattern = "*overview.txt"){
    files = list.files(path, pattern, full.names = TRUE)
    lapply(files, read_delim, delim="\t") %&gt;% bind_rows()
  }
  dbcan_df&lt;-lapply_read_delim_bind_rows(ruta_dbcan)
  # Reading data ----------------------------------------------------------####
  dbcan_df_format&lt;- suppressWarnings(
    dbcan_df %&gt;%
      filter( .data$`#ofTools` &gt;1) %&gt;%
      rename(Bin_name=.data$`Gene ID`) %&gt;%
      mutate( hmmer2=str_replace_all(.data$HMMER, "[[:punct:]]",  "\t")) %&gt;%
      separate(.data$hmmer2, c("dbNamesHMM"),  sep="\t") %&gt;%
      mutate(hotpep2=str_replace_all(.data$Hotpep, "[[:punct:]]", "\t")) %&gt;%
      separate(.data$hotpep2, c("dbNameshotpep"), sep="\t") %&gt;%
      mutate(diamond2=str_replace_all(.data$DIAMOND, "[[:punct:]]", "\t")) %&gt;%
      separate(.data$diamond2, c("dbNamesdiamond"), sep="\t") %&gt;%
      unite("dbCAN_names", .data$dbNamesHMM, .data$dbNameshotpep,  .data$dbNamesdiamond, 
           sep="_", remove = F) %&gt;%  
      mutate(dbCAN_names=str_replace_all(.data$dbCAN_names, "^_", "")) %&gt;%
      separate(.data$dbCAN_names, c("dbCAN_names"), sep="_") 
```
---

## Funcion de lectura de archivo


```r
%&gt;% mutate(dbCAN = case_when(str_detect(.data$dbCAN_names, "CBM") ~ 
                                "carbohydrate-binding module [CBM]",
                              str_detect(.data$dbCAN_names, "CE") ~ 
                                "carbohydrate esterases [CEs]",
                              str_detect(.data$dbCAN_names, "GH") ~ 
                                "glycoside hydrolases [GHs]",
                              str_detect(.data$dbCAN_names, "GT") ~ 
                                "glycosyltransferases [GTs]",
                              str_detect(.data$dbCAN_names, "PL") ~ 
                                "polysaccharide lyases [PLs]")) %&gt;%
      mutate_if(is.character, str_trim) %&gt;%
    dplyr::select(.data$Bin_name, .data$dbCAN_names) %&gt;%
    calc_abundance(analysis = "dbCAN") %&gt;% 
    dplyr::select(-.data$Scaffold_name))
  # Menssage --------------------------------------------------------------####
  initial&lt;-dim(dbcan_df)
  final&lt;-dim(dbcan_df %&gt;%
               filter( .data$`#ofTools` &gt;1))
  signals&lt;- dbcan_df %&gt;% group_by(.data$Signalp) %&gt;% count()

  print(paste0("Input Genes = " , initial[1]))
  print(paste0("Output filtered Genes = " , final[1]))
  print(paste0("Percentage remained = " , round(final[1]/initial[1]*100), "%"))
  print(paste0("Number of Genes with no signals = " , signals[2]))

  return(dbcan_df_format)
}
```



---
## Probando la función


```r
data_dbcan&lt;-"../Data/"
table_test&lt;-read_dbcan(data_dbcan)
```

```
## [1] "Input Genes = 1206"
## [1] "Output filtered Genes = 265"
## [1] "Percentage remained = 22%"
## [1] "Number of Genes with no signals = 1206"
```

```r
head(table_test)
```

```
## # A tibble: 6 × 3
##   Bin_name         dbCAN Abundance
##   &lt;chr&gt;            &lt;chr&gt;     &lt;int&gt;
## 1 Meg22_46_Bin_204 GH51          1
## 2 Meg22_46_Bin_204 GT2           3
## 3 Meg22_46_Bin_204 PL10          1
## 4 Meg22_46_Bin_204 GH32          1
## 5 Meg22_46_Bin_204 GH141         4
## 6 Meg22_46_Bin_204 GH36          1
```
---

# Funcion para la visualización de datos en bubble plot

[Liga a la función](https://github.com/Steph0522/Rbims_workflow/blob/main/R/2022_08_05-PLOT-bubble_dbcan.R)


```r
bubble_dbcan&lt;-function(tibble_dbcan,
                    x_axis, 
                    y_axis,
                    data_experiment=NULL,
                    color_character=NULL,
                    order_bins=NULL,
                    order_metabolism=NULL,
                    color_pallet=NULL, 
                    range_size=NULL,
                    x_labs=TRUE,
                    y_labs=TRUE,
                    text_x=NULL,
                    text_y=NULL){
 ..........
  return(plot_bubble)
}
```
---

# Corriendo la función de bubble plot




```r
Metadatos&lt;- data.frame(
  Bin_name = unique(table_test$Bin_name),
  Specie = c("Specie_a", "Specie_b", "Specie_c",
             "Specie_d", "Specie_e", "Specie_f")) 
bubble_dbcan(tibble_dbcan = table_test,
            x_axis = Bin_name, 
            y_axis = dbCAN,
            range_size = c(1,10),
            data_experiment = Metadatos,
            color_character = Specie,
            y_labs=FALSE,
            x_labs=FALSE)  
```

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
